Subject: [PATCH] 11_kafka_integration_3_HWD
---
Index: src/test/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepositoryTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepositoryTest.java b/src/test/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepositoryTest.java
--- a/src/test/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepositoryTest.java	(revision aeb0c7051e5657d4baf008aa05750a6a4b6dbfc9)
+++ b/src/test/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepositoryTest.java	(date 1731252293682)
@@ -13,7 +13,9 @@
 import ru.javaops.cloudjava.ordersservice.BaseTest;
 import ru.javaops.cloudjava.ordersservice.config.R2dbcConfig;
 import ru.javaops.cloudjava.ordersservice.storage.model.MenuOrder;
+import ru.javaops.cloudjava.ordersservice.storage.model.OrderStatus;
 
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
 import static ru.javaops.cloudjava.ordersservice.testdata.TestConstants.*;
 
 @Import({R2dbcConfig.class})
@@ -43,7 +45,7 @@
     @Test
     void findAllByCreatedBy_returnsCorrectSortedByDateAsc() {
         var pageRequest = PageRequest.of(0, 2)
-                .withSort(Sort.by(Sort.Direction.ASC,"createdAt"));
+                .withSort(Sort.by(Sort.Direction.ASC, "createdAt"));
         Flux<MenuOrder> orders = repository.findAllByCreatedBy(USERNAME_ONE, pageRequest);
         StepVerifier.create(orders)
                 .expectNextMatches(order ->
@@ -58,10 +60,30 @@
     @Test
     void findAllByCreatedBy_returnsEmptyListWhenUserHasNoOrders() {
         var pageRequest = PageRequest.of(0, 10)
-                .withSort(Sort.by(Sort.Direction.ASC,"createdAt"));
+                .withSort(Sort.by(Sort.Direction.ASC, "createdAt"));
         Flux<MenuOrder> orders = repository.findAllByCreatedBy("unknown user", pageRequest);
         StepVerifier.create(orders)
                 .expectNextCount(0)
                 .verifyComplete();
     }
+
+    @Test
+    void updateStatusById_updatesStatusOfExistingOrder() {
+        var order = repository.findAll().blockFirst();
+        var orderId = order.getId();
+
+        repository.updateStatusById(orderId, OrderStatus.ACCEPTED).block();
+        var updated = repository.findById(orderId).block();
+        assertThat(updated.getStatus()).isEqualTo(OrderStatus.ACCEPTED);
+        assertThat(updated.getUpdatedAt()).isAfter(order.getUpdatedAt());
+    }
+
+    @Test
+    void updateStatusById_doesNothingIfOrderNotExists() {
+        Long orderId = 1000L;
+
+        repository.updateStatusById(orderId, OrderStatus.ACCEPTED).block();
+        var updated = repository.findById(orderId).block();
+        assertThat(updated).isNull();
+    }
 }
Index: src/main/java/ru/javaops/cloudjava/ordersservice/kafka/KafkaOrderDispatchListener.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/ru/javaops/cloudjava/ordersservice/kafka/KafkaOrderDispatchListener.java b/src/main/java/ru/javaops/cloudjava/ordersservice/kafka/KafkaOrderDispatchListener.java
--- a/src/main/java/ru/javaops/cloudjava/ordersservice/kafka/KafkaOrderDispatchListener.java	(revision aeb0c7051e5657d4baf008aa05750a6a4b6dbfc9)
+++ b/src/main/java/ru/javaops/cloudjava/ordersservice/kafka/KafkaOrderDispatchListener.java	(date 1731264696607)
@@ -2,17 +2,44 @@
 
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.kafka.support.Acknowledgment;
+import org.springframework.kafka.support.KafkaHeaders;
+import org.springframework.messaging.handler.annotation.Header;
 import org.springframework.stereotype.Component;
+import ru.javaops.cloudjava.OrderDispatchedEvent;
+import ru.javaops.cloudjava.ordersservice.storage.model.OrderStatus;
 import ru.javaops.cloudjava.ordersservice.storage.repositories.MenuOrderRepository;
 
+import java.time.Duration;
+
 @Slf4j
 @Component
 @RequiredArgsConstructor
 public class KafkaOrderDispatchListener {
 
     private final MenuOrderRepository menuOrderRepository;
+    @Value("${kafkaprops.nack-sleep-duration}")
+    private Duration nackSleepDuration;
 
-    // TODO
-    public void consumeOrderDispatchEvent() {
+    @KafkaListener(topics = {"${kafkaprops.order-dispatch-topic}"})
+    public void consumeOrderDispatchEvent(OrderDispatchedEvent event,
+                                          @Header(KafkaHeaders.RECEIVED_KEY) String key,
+                                          @Header(KafkaHeaders.RECEIVED_PARTITION) Integer partition,
+                                          @Header(KafkaHeaders.RECEIVED_TOPIC) String topic,
+                                          Acknowledgment acknowledgment) {
+        log.info("Received OrderDispatchedEvent from Kafka: {}. Key: {}. Partition: {}. Topic: {}", event, key, partition, topic);
+        try {
+            menuOrderRepository
+                    .updateStatusById(event.getOrderId(), OrderStatus.fromString(event.getStatus().name()))
+                    .block();
+            log.info("Successfully updated OrderStatus to {} for order with ID={}", event.getStatus(), event.getOrderId());
+            acknowledgment.acknowledge();
+        } catch (Exception e) {
+            log.error("Failed to update OrderStatus to {} for order with ID={}", event.getStatus(), event.getOrderId());
+            // don't acknowledge
+            acknowledgment.nack(nackSleepDuration);
+        }
     }
 }
\ No newline at end of file
Index: src/main/resources/application.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/resources/application.yml b/src/main/resources/application.yml
--- a/src/main/resources/application.yml	(revision aeb0c7051e5657d4baf008aa05750a6a4b6dbfc9)
+++ b/src/main/resources/application.yml	(date 1731266794559)
@@ -42,9 +42,27 @@
         max-interval: 2000
         multiplier: 1.1
 
+  kafka:
+    bootstrap-servers: localhost:9097
+    consumer:
+      group-id: ${spring.application.name}
+      key-deserializer: org.apache.kafka.common.serialization.StringDeserializer
+      value-deserializer: io.confluent.kafka.serializers.KafkaAvroDeserializer
+      properties:
+        specific.avro.reader: true
+        schema.registry.url: http://localhost:8081
+      auto-offset-reset: earliest
+      enable-auto-commit: false
+    listener:
+      ack-mode: manual
+
 configserver:
   import: optional:configserver:${CONFIG_SERVER_URL:http://localhost:9095}
 
+kafkaprops:
+  order-dispatch-topic: v1.orders_dispatch
+  nack-sleep-duration: 100ms
+
 external:
   # url для связи с Menu Service
   menu-service-url: "http://localhost:9091"
Index: src/main/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepository.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepository.java b/src/main/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepository.java
--- a/src/main/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepository.java	(revision aeb0c7051e5657d4baf008aa05750a6a4b6dbfc9)
+++ b/src/main/java/ru/javaops/cloudjava/ordersservice/storage/repositories/MenuOrderRepository.java	(date 1731251348891)
@@ -1,11 +1,21 @@
 package ru.javaops.cloudjava.ordersservice.storage.repositories;
 
 import org.springframework.data.domain.Pageable;
+import org.springframework.data.r2dbc.repository.Modifying;
+import org.springframework.data.r2dbc.repository.Query;
 import org.springframework.data.repository.reactive.ReactiveCrudRepository;
+import org.springframework.transaction.annotation.Transactional;
 import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
 import ru.javaops.cloudjava.ordersservice.storage.model.MenuOrder;
+import ru.javaops.cloudjava.ordersservice.storage.model.OrderStatus;
 
 public interface MenuOrderRepository extends ReactiveCrudRepository<MenuOrder, Long> {
 
     Flux<MenuOrder> findAllByCreatedBy(String username, Pageable pageable);
+
+    @Transactional
+    @Modifying
+    @Query("UPDATE orders SET status = :newStatus, updated_at = CURRENT_TIMESTAMP WHERE id = :orderId")
+    Mono<Void> updateStatusById(Long orderId, OrderStatus newStatus);
 }
\ No newline at end of file
